specs/derivation.md

```diff
@@ -974,46 +978,155 @@ Some network upgrades require automated contract changes or deployments at speci
-- `to`: TBD (depending on multi-call approach)
+- `from`: `0x4210000000000000000000000000000000000001`
+- `to`: `null`,
 - `mint`: `0`
 - `value`: `0`
-- `gasLimit`: gas limit TBD
-- `isCreation`: `false`
-- `data`: bytecode TBD, a multi-call:
-  - deploy new contract implementation
-  - upgrade of `L1Block` through `ProxyAdmin`
-- `sourceHash`: `0x7dc74874297a8937186fdbec57ad344647a522de456088557e5fdeda88f66ddd`,
-  computed with the "Upgrade-deposited" type, with `intent = "Eclipse: L1Block upgrade"`
+- `gasLimit`: `1,000,000` (TBC)
+- `data`: GasPriceOracle deploy bytecode (TBC when PR for contracts is merged)
+- `sourceHash`: `0xa312b4510adf943510f05fcc8f15f86995a5066bd83ce11384688ae20e6ecf42`,
+  computed with the "Upgrade-deposited" type, with `intent = "Ecotone: Gas Price Oracle Deployment"
+
+This results in the Ecotone GasPriceOracle contract being deployed to `0xb528D11cC114E026F138fE568744c6D45ce6Da7A`, to verify:
+
+```bash
+cast compute-address --nonce=0 0x4210000000000000000000000000000000000001
+Computed Address: 0xb528D11cC114E026F138fE568744c6D45ce6Da7A
+```
 
 Verify `sourceHash`:
 
 ```bash
 # compute intent hash:
-cast keccak "Eclipse: L1Block upgrade"
-# 0x831b745c7397f93704ae55eb0100bf3c56fe9e304d3f21c1a93ec25f736fea26
+cast keccak "Ecotone: Gas Price Oracle Deployment"
+# 0x529c922cad7984deae1f8311397f61998c2d101c184693006c91384c0faae8be
+
+# source hash type:
+# 0x0000000000000000000000000000000000000000000000000000000000000002
+
+# compute source hash:
+cast keccak 0x0000000000000000000000000000000000000000000000000000000000000002529c922cad7984deae1f8311397f61998c2d101c184693006c91384c0faae8be
+# 0xa312b4510adf943510f05fcc8f15f86995a5066bd83ce11384688ae20e6ecf42
+```
+
+##### L1Block Proxy Update
+
+This transaction updates the L1Block proxy to point to the new L1Block implementation.
+
+A deposit transaction is derived with the following attributes:
+
+- `from`: `Proxy Admin`
+- `to`: `L1 Block Proxy`,
+- `mint`: `0`
+- `value`: `0`
+- `gasLimit`: `1,000,000` (TBC)
+- `data`: `upgradeTo(0x07dbe8500fc591d1852B76feE44d5a05e13097Ff)`
+- `sourceHash`: `0x18acb38c5ff1c238a7460ebc1b421fa49ec4874bdf1e0a530d234104e5e67dbc`,
+  computed with the "Upgrade-deposited" type, with `intent = "Ecotone: L1 Block Proxy Update"
+
+```bash
+# compute intent hash:
+cast keccak "Ecotone: L1 Block Proxy Update"
+# 0x8eb5485869caa58625c9c5f54152d8f25133495d52f3cd99861072299010d005
+
+# source hash type:
+# 0x0000000000000000000000000000000000000000000000000000000000000002
+
+# compute source hash:
+cast keccak 0x00000000000000000000000000000000000000000000000000000000000000028eb5485869caa58625c9c5f54152d8f25133495d52f3cd99861072299010d005
+# 0x18acb38c5ff1c238a7460ebc1b421fa49ec4874bdf1e0a530d234104e5e67dbc
+```
+
+##### GasPriceOracle Proxy Update
+
+This transaction updates the GasPriceOracle proxy to point to the new GasPriceOracle implementation.
+
+A deposit transaction is derived with the following attributes:
+
+- `from`: `Proxy Admin`
```

Comment:
This is an interesting because if the spec says that from should be the `ProxyAdmin` it is possible to configure the L2 in a way that breaks the hardfork by altering the 1967 admin of the GasPriceOracle at the application layer. Because of this reason, I don't think its a good idea to make the from be the `ProxyAdmin` in the spec. It would be better to make the spec say that the `from` should be the address in the 1967 admin slot. Altho I would be worried about that creating an invalid tx because of [this](https://github.com/ethereum/go-ethereum/blob/8c2d455ccd216fb8589c15339392ce9640d8090d/core/state_transition.go#L285) check

Reply:
To confirm I understand, there's two issues here:

1. These txns should originate from an EOA.
2. A chain operator could change the admin of the predployed proxies from `ProxyAdmin` to another address.

For 1) I'm not super familiar with `op-geth`, but I think these checks [maybe skipped for deposit txns](https://github.com/ethereum-optimism/op-geth/blob/8e15470125ba3689dd7288e2aa7bbdca8730701b/core/state_transition.go#L287-L302). However the comment "no need to check if EOA (L1 already verified it for us)" sounds like we maybe misusing this.

For 2) I think having the depositor address have sudo access to the predeployed proxies makes this a lot cleaner for this upgrade + future upgrades vs. resolving owners etc.

Do you think it's worthwhile in this hardfork changing the proxy implementations to include this?

Reply:
You are right that 1) is a non issue because of the implementation of how deposit transactions work. This is great and makes our lives way easier!

Re 2) I do believe it would be much cleaner to make this change although it will add scope. I can support with implementation but ultimately going to lean on you to determine if you think it is something that we can do in time. It would involve some additional system txs as well as 2 additional contracts that would need to go through an audit

Reply:
Ok we have agreed that we want `from` to be `address(0)`: https://github.com/ethereum-optimism/optimism/blob/e57ad63a23619e5da880c5140e33948d3e765dc5/packages/contracts-bedrock/src/universal/Proxy.sol#L27